core/domain/stats_jobs_one_off.py

```diff
@@ -18,14 +18,403 @@
+                            completed.
+                        'num_actual_starts': int. # of times exploration was
+                            actually started.
+                        'state_stats_mapping': A dict containing the values of
+                            stats for the states of the exploration. It is
+                            formatted as follows:
+                            {
+                                state_name: {
+                                    'total_answers_count',
+                                    'useful_feedback_count',
+                                    'total_hit_count',
+                                    'first_hit_count',
+                                    'num_completions'
+                                }
+                            }
+                    }
+        """
+        reduced_state_stats_mapping = {
+            state_name: {
+                'total_answers_count': item.state_stats_mapping[state_name][
+                    'total_answers_count_v1'],
+                'useful_feedback_count': item.state_stats_mapping[state_name][
+                    'useful_feedback_count_v1'],
+                'total_hit_count': item.state_stats_mapping[state_name][
+                    'total_hit_count_v1'],
+                'first_hit_count': item.state_stats_mapping[state_name][
+                    'first_hit_count_v1'],
+                'num_completions': item.state_stats_mapping[state_name][
+                    'num_completions_v1']
+            } for state_name in item.state_stats_mapping
+        }
+
+        yield(
+            item.exp_id,
+            {
+                'version': item.exp_version,
+                'num_starts': item.num_starts_v1,
+                'num_completions': item.num_completions_v1,
+                'num_actual_starts': item.num_actual_starts_v1,
+                'state_stats_mapping': reduced_state_stats_mapping
+            })
+
+    @staticmethod
+    def reduce(key, stringified_values):
+        for value in stringified_values:
+            value = ast.literal_eval(value)
+            version = value['version']
+
+            num_starts = value['num_starts']
+            num_completions = value['num_completions'],
+            num_actual_starts = value['num_actual_starts']
+
+            StatisticsAuditV2.require_non_negative(key, 'num_starts', value)
+            StatisticsAuditV2.require_non_negative(
+                key, 'num_completions', value)
+            StatisticsAuditV2.require_non_negative(
+                key, 'num_actual_starts', value)
+
+            # Number of starts must never be less than the number of
+            # completions.
+            if num_starts < num_completions:
+                yield ('Completions > starts: exp_id:%s version:%s %s>%s' % (
+                    key, version, num_completions, num_starts),)
+
+            # Number of actual starts must never be less than the number of
```

Comment:
Some explorations have one state (rare, but possible). How will you handle those?

(I'm not sure if this is the case for _published_ explorations. Might be worth checking whether it's possible to publish a 1-state exploration on the dev server. Also, might be worth prohibiting this going forward; a one-state exploration is unlikely to be of much use.)

Reply:
But then such an exploration may change over time to have more than one state, and more generally there may be occasional periods in an exploration's history where it has one state and where it has more than one. Maybe just keeping num_actual_starts at 0 is the consistent thing to do?